src/Type/Constant/ConstantArrayTypeBuilder.php

```diff
@@ -162,7 +162,7 @@ public function setOffsetValueType(?Type $offsetType, Type $valueType, bool $opt
 				$this->keyTypes[] = $offsetType;
 				$this->valueTypes[] = $valueType;
 
-				if ($offsetType instanceof ConstantIntegerType) {
+				if ($offsetType instanceof ConstantIntegerType && $offsetType->getValue() >= 0) {
```

Comment:
I'm worried what happens there because suddenly the `nextAutoIndexes` are not updated for negative numbers.

But they should be updated. The behaviour even changed on PHP 8.3+: https://3v4l.org/DWBH9